### PR TITLE
Support global atom_* atomic functions for 32-bit integer types

### PR DESCRIPTION
**Pull request by kpet**
All these pass the OpenCL conformance tests run via clvk on
NVidia hardware.

Signed-off-by: Kévin Petit <kpet@free.fr>

### CODE REVIEW

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 1]
```diff
@@ -1950,6 +1956,22 @@ bool ReplaceOpenCLBuiltinPass::replaceAtomics(Module &M) {
   }
 
   const std::map<const char *, llvm::AtomicRMWInst::BinOp> Map2 = {
+      {"_Z8atom_addPU3AS1Vii", llvm::AtomicRMWInst::Add},
+      {"_Z8atom_addPU3AS1Vjj", llvm::AtomicRMWInst::Add},
+      {"_Z8atom_subPU3AS1Vii", llvm::AtomicRMWInst::Sub},
+      {"_Z8atom_subPU3AS1Vjj", llvm::AtomicRMWInst::Sub},
+      {"_Z9atom_xchgPU3AS1Vii", llvm::AtomicRMWInst::Xchg},
+      {"_Z9atom_xchgPU3AS1Vjj", llvm::AtomicRMWInst::Xchg},
+      {"_Z8atom_minPU3AS1Vii", llvm::AtomicRMWInst::Min},
```
Comment (alan-baker) [inline lib/ReplaceOpenCLBuiltinPass.cpp]: One of these should map to AtomicRMWInst::UMin. Probably the `j` version, but I don't remember all the mangling off hand.

File: lib/ReplaceOpenCLBuiltinPass.cpp [hunk 2]
```diff
@@ -1950,6 +1956,22 @@ bool ReplaceOpenCLBuiltinPass::replaceAtomics(Module &M) {
...
+      {"_Z8atom_minPU3AS1Vjj", llvm::AtomicRMWInst::Min},
+      {"_Z8atom_maxPU3AS1Vii", llvm::AtomicRMWInst::Max},
```
Comment (alan-baker) [inline lib/ReplaceOpenCLBuiltinPass.cpp]: Same here, but AtomicRMWInst::UMax.

File: test/AtomicBuiltins/atom_max_uint.cl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// CHECK-DAG: %[[uint_1:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 1
+// CHECK-DAG: %[[uint_80:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 80
+// CHECK-DAG: %[[uint_42:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 42
+// CHECK-DAG: %[[__original_id_12]] = OpVariable %[[_ptr_StorageBuffer__struct_3]] StorageBuffer
+// CHECK-DAG: %[[__original_id_13]] = OpVariable %[[_ptr_StorageBuffer__struct_3]] StorageBuffer
+// CHECK:     %[[__original_id_14]] = OpFunction %[[void]] None %[[__original_id_6]]
+// CHECK:     %[[__original_id_15:[0-9]+]] = OpLabel
+// CHECK:     %[[__original_id_16:[0-9]+]] = OpAccessChain %[[_ptr_StorageBuffer_uint]] %[[__original_id_12]] %[[uint_0]] %[[uint_0]]
+// CHECK:     %[[__original_id_17:[0-9]+]] = OpAccessChain %[[_ptr_StorageBuffer_uint]] %[[__original_id_13]] %[[uint_0]] %[[uint_0]]
+// CHECK:     %[[__original_id_18:[0-9]+]] = OpAtomicSMax %[[uint]] %[[__original_id_17]] %[[uint_1]] %[[uint_80]] %[[uint_42]]
```
Comment (alan-baker) [inline test/AtomicBuiltins/atom_max_uint.cl]: This should end up changing.

File: test/AtomicBuiltins/atom_min_uint.cl [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+// CHECK-DAG: %[[uint_1:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 1
+// CHECK-DAG: %[[uint_80:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 80
+// CHECK-DAG: %[[uint_42:[0-9a-zA-Z_]+]] = OpConstant %[[uint]] 42
+// CHECK-DAG: %[[__original_id_12]] = OpVariable %[[_ptr_StorageBuffer__struct_3]] StorageBuffer
+// CHECK-DAG: %[[__original_id_13]] = OpVariable %[[_ptr_StorageBuffer__struct_3]] StorageBuffer
+// CHECK:     %[[__original_id_14]] = OpFunction %[[void]] None %[[__original_id_6]]
+// CHECK:     %[[__original_id_15:[0-9]+]] = OpLabel
+// CHECK:     %[[__original_id_16:[0-9]+]] = OpAccessChain %[[_ptr_StorageBuffer_uint]] %[[__original_id_12]] %[[uint_0]] %[[uint_0]]
+// CHECK:     %[[__original_id_17:[0-9]+]] = OpAccessChain %[[_ptr_StorageBuffer_uint]] %[[__original_id_13]] %[[uint_0]] %[[uint_0]]
+// CHECK:     %[[__original_id_18:[0-9]+]] = OpAtomicSMin %[[uint]] %[[__original_id_17]] %[[uint_1]] %[[uint_80]] %[[uint_42]]
```
Comment (alan-baker) [inline test/AtomicBuiltins/atom_min_uint.cl]: This too.

File: lib/ReplaceOpenCLBuiltinPass.cpp (hunk 1 above)
Comment (kpet) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: Yes, you're right, well spotted. The conformance tests probably need to be fixed as well.

File: lib/ReplaceOpenCLBuiltinPass.cpp (hunk 2 above)
Comment (kpet) [reply lib/ReplaceOpenCLBuiltinPass.cpp]: Done.

**Comment by kpet** [comment]
@alan-baker Thanks for the comments. I've updated the PR. Looking at the conformance tests, they're not that easy to fix without adding even more to the existing (already pretty bad) code duplication. I've raised https://github.com/KhronosGroup/OpenCL-CTS/issues/10 so that it gets attention from the working group.

**Pull request merged**